src/qibo/gates/gates.py

```diff
@@ -2,11 +2,89 @@
 from typing import List
 
 import numpy as np
+import sympy as sp
 
 from qibo.config import PRECISION_TOL, raise_error
 from qibo.gates.abstract import Gate, ParametrizedGate
 
 
+class Parameter:
+    """Object which allows complex gate parameters. Several trainable parameter
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter"""
+
+    def __init__(self, func, trainablep, featurep=None):
+        self._trainablep = trainablep
+        self._featurep = featurep
+        self.nparams = len(trainablep)
+
+        if isinstance(featurep, list):
+            self.nfeat = len(featurep)
+        else:
+            self.nfeat = 0
+        self.lambdaf = func
+        self.derivatives = self.calculate_derivatives()
+
+    def _apply_func(self, function, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+        if self._featurep is not None:
+            if isinstance(self._featurep, list):
+                params.extend(self._featurep)
+            else:
+                params.append(self._featurep)
+        if fixed_params:
+            params.extend(fixed_params)
+        else:
+            params.extend(self._trainablep)
+        return float(function(*params))
+
+    def _update_params(self, trainablep=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if trainablep is not None:
+            self._trainablep = trainablep
+        if feature and self._featurep:
+            self._featurep = feature
+
+    def get_params(self, trainablep=None, feature=None):
```

Comment:
Same here, I would prefer
```suggestion
    def get_parameters(self, trainablep=None, feature=None):
```
particularly because this is exposed to the user.